### PR TITLE
Proficiency routing

### LINKED ISSUE #195: Effectiveness/Proficiency Routing
Rather than attempting to route based on category, update category routing or replace it with "rating" routing. Rating routing would take the statistics from sources such as the one below and route from the best to the worst model.
https://benchmarks.llmonitor.com/

### PR DESCRIPTION
**Pull request by BallenAlain**
resolves #195 

notes to consider:
- data taken from https://huggingface.co/spaces/lmsys/chatbot-arena-leaderboard
- no data for vertex AI models except gemini and cohere
- sort models based on elo rating (model with highest elo rating will be chosen)
- replaced routing type from category to proficiency in llmproxy init unit tests since category routing will be deprecated soon

elo rating data notes:
- GPT-4 (points to gpt-4-0613 see https://platform.openai.com/docs/models/gpt-4-and-gpt-4-turbo)
- No data found online for gpt-3.5-turbo-0613, model will be deprecated https://ai.stackexchange.com/questions/40111/openai-what-is-the-difference-between-model-gpt-3-5-turbo-and-gpt-3-5-turbo
- Mistral-7b-v0.1 uses elo rating for mistral-7b-instruct-v0.1
- command-r is similar to mixtral 8x7b but ranks slightly higher. Will have a similar elo rating (see https://huggingface.co/spaces/HuggingFaceH4/open_llm_leaderboard )
- text-bison and chat-bison will give the same quality of results but text-bison seems more applicable for our application and should rank slightly higher than chat-bison
- Palm and codey models use the same model but each is optimized for different tasks. Palm models will rank higher than codey models since it is more applicable 
- For codey models, i think ranking should be as follows from most important to least: code completion (code-gecko), codechat bison, code completion (code-bison)
- see code comments for command, command-light, command nightly, command-light-nightly (https://docs.cohere.com/docs/command-beta)

### CODE REVIEW

File: proxyllm/proxyllm.py [hunk 1]
```diff
@@ -96,7 +97,6 @@ def _setup_available_models(settings: List[Dict[str, Any]]) -> Dict[str, Any]:
 
             # Loop through and aggregate all of the variations of "models" of each provider
             provider_models = set()
-            model_costs = {}
```
Comment (BallenAlain) [inline proxyllm/proxyllm.py:99]: not being used anywhere

File: proxyllm/proxyllm.py [hunk 2]
```diff
@@ -219,7 +219,7 @@ def _setup_user_models(
         ) from e
 
 
-def _setup_models_cost_data(settings: List[Dict[str, Any]]) -> Dict[str, Any]:
+def _setup_model_data(settings: List[Dict[str, Any]]) -> Dict[str, Any]:
```
Comment (BallenAlain) [inline proxyllm/proxyllm.py:222]: repurposed function to get model data from yml for both cost and proficiency routing

File: proxyllm/proxyllm.py [hunk 3]
```diff
@@ -566,3 +574,91 @@ def _category_route(self, prompt: str):
...
+
+        Args:
+            prompt (str): The input prompt to generate text for.
+
+        Returns:
+            CompletionResponse: The generated text response along with any errors encountered.
+        """
+
+        min_heap = MinHeap()
+        logger.log(msg="Sorting fetched models based on elo rating...", color="GEEN")
```
Comment (laxitshahi) [inline proxyllm/proxyllm.py:590]: typro

**Review by laxitshahi — CHANGES_REQUESTED** [review]
I will look at this review again after claude has been merged in

File: proxyllm/proxyllm.py (hunk 3 above)
Comment (BallenAlain) [reply proxyllm/proxyllm.py]: fixed

File: proxyllm/proxyllm.py [hunk 4]
```diff
@@ -501,7 +509,7 @@ def _category_route(self, prompt: str):
             )
 
             logger.log(
-                msg="Sorting fetched models based on proficency...",
+                msg="Sorting fetched models based on proficiency...",
             )
```
[the remark this answers is not in the document]
Comment (BallenAlain) [reply proxyllm/proxyllm.py:513]: relative ranking? elo is basically ranking but relative to each other

**Review by laxitshahi — CHANGES_REQUESTED** [review]
You need to add the route type to the literal:
image
To avoid this error:
image

--
claude was merged in, so you will have to add that elo for that and make sure the routing makes sense

**Comment by BallenAlain** [comment]
> You need to add the route type to the literal: image To avoid this error: image
> 
> -- claude was merged in, so you will have to add that elo for that and make sure the routing makes sense

done

File: proxyllm/proxyllm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (BallenAlain) [reply proxyllm/proxyllm.py:513]: renamed to elo

File: proxyllm/proxyllm.py [hunk 5]
```diff
@@ -566,3 +574,91 @@ def _category_route(self, prompt: str):
...
+                logger.log(
+                    level="ERROR",
+                    msg=f"Error when making request to model: {e}",
+                )
+
+                logger.log(level="ERROR", msg="(•᷄ ∩ •᷅)", file_logger_on=False)
+
+                logger.log(
+                    level="ERROR",
+                    msg="========PROFICIENCY ROUTING FAILED!===========\n",
```
[the remark this answers is not in the document]
Comment (BallenAlain) [reply proxyllm/proxyllm.py]: will address in another PR (deprecating category routing)

File: proxyllm/config/internal_config.py [hunk 6]
```diff
@@ -29,54 +30,64 @@
                 "name": "gpt-4-0125-preview",
                 "cost_per_token_input": 10e-06,
                 "cost_per_token_output": 30e-06,
+                "elo": 1248,
             },
             {
                 "name": "gpt-4-1106-preview",
```
[the remark this answers is not in the document]
Comment (BallenAlain) [reply proxyllm/config/internal_config.py:36]: done

File: proxyllm/config/internal_config.py [hunk 7]
```diff
@@ -206,31 +237,37 @@
...
                 "name": "gemini-pro",
                 "cost_per_token_input": 1.25e-07,
                 "cost_per_token_output": 3.75e-07,
+                "elo": 1125,
             },
             {
                 "name": "chat-bison",
                 "cost_per_token_input": 1.5e-06,
                 "cost_per_token_output": 2e-06,
+                "elo": 0,
```
[the remark this answers is not in the document]
Comment (BallenAlain) [reply proxyllm/config/internal_config.py]: see comments and pr descriptino

**Review by laxitshahi — APPROVED** [review]

**Pull request merged**